### PR TITLE
chore: watcher cyclical dependencies

### PR DESCRIPTION
**Pull request by jonathanyeung**
## chore: various cyclical dependencies
Fixing cyclical dependencies in the following areas:
- testUtils
- workspaceInitializer
- windowWatcher
- commandRegistrar

Some of the changes include
- splitting out component/utils.ts partially to a new NotePickerUtils.ts
- Moving onFirstOpen logic from windowWatcher to workspaceWatcher
- Creating a MockDendronExtension that implements IDendronExtension to help with our testing.

After this change, madge reports 68 remaining cyclical dependencies in plugin-core

---

# Pull Request Checklist

If this is your first time submitting a pull request to Dendron, copy and paste the full [Dendron Review Checklist](https://docs.dendron.so/notes/1EoNIXzgmhgagqcAo9nDn.html) into this request and check off each item as necessary. 

This template contains the short checklist which is used by the Dendron core team. 

## Testing
- [ ] [Write Tests](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#writing-tests) 
- [ ] [Confirm existing tests pass](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#executing-tests)
- [ ] [Confirm manual testing](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#manual-testing) 
- [ ] If your tests changes an existing snapshot, make sure that snapshots are [updated](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#updating-test-snapshots)
- [ ] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), make sure that it is included in the [test workspace](https://docs.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)
### Docs
- [ ] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
### Analytics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://docs.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### CODE REVIEW

File: packages/plugin-core/src/test/testUtilsv2.ts [hunk 1]
```diff
@@ -472,4 +471,15 @@ export const stubWorkspace = ({ wsRoot, vaults }: WorkspaceOpts) => {
   stubWorkspaceFolders(wsRoot, vaults);
 };
 
+/**
+ *  Releases all registered VS Code Extension resouces such as commands and
+ *  providers
+ * @param ctx
+ */
+export function cleanupVSCodeContextSubscriptions(ctx: ExtensionContext) {
```
Comment (hikchoi) [inline packages/plugin-core/src/test/testUtilsv2.ts:479]: Is there a reason this was moved here?

File: packages/plugin-core/src/test/MockDendronExtension.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  get noteSyncService(): INoteSyncService {
+    throw new Error("Method not implemented");
+  }
+  get lookupControllerFactory(): ILookupControllerV3Factory {
+    throw new Error("Method not implemented");
+  }
+  get noteLookupProviderFactory(): INoteLookupProviderFactory {
+    throw new Error("Method not implemented");
+  }
+  // schemaLookupProviderFactory: ISchemaLookupProviderFactory;
```
Comment (hikchoi) [inline packages/plugin-core/src/test/MockDendronExtension.ts:77]: remove.

File: packages/plugin-core/src/components/lookup/NotePickerUtils.ts [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  SchemaUtils,
+} from "@dendronhq/common-all";
+import { DLogger, getDurationMilliseconds } from "@dendronhq/common-server";
+import { HistoryService, LinkUtils } from "@dendronhq/engine-server";
+import _ from "lodash";
+import path from "path";
+import { ExtensionProvider } from "../../ExtensionProvider";
+import { Logger } from "../../logger";
+import { VSCodeUtils } from "../../vsCodeUtils";
+import { getDWorkspace } from "../../workspace";
```
Comment (hikchoi) [inline packages/plugin-core/src/components/lookup/NotePickerUtils.ts:20]: do these touch on too many places to swap out in this PR? if so we should take note of it and clean it up on the next round of cleanup.

**Review by hikchoi — COMMENTED** [review]
I see a few bare `getExtension()` and `getDWorkspace()` in the files that were touched. were these out of the scope of this round?

Other than that left a few nitpicks.

a doc on how to use the `MockDendronExtension` would be great!

File: packages/plugin-core/src/components/lookup/NotePickerUtils.ts (hunk 3 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/components/lookup/NotePickerUtils.ts:20]: Nope - just missed it.  Cleaned up in latest commit.

**Comment by jonathanyeung** [comment]
Thanks for the review @hikchoi. 

> a doc on how to use the MockDendronExtension would be great!

Will find the right place to add, and then post team. It'll be required more frequently as we have more constructor signatures with `IDendronExtension`

File: packages/plugin-core/src/commands/ConvertLink.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ import {
 } from "@dendronhq/engine-server";
 import _ from "lodash";
 import { LookupControllerV3CreateOpts } from "../components/lookup/LookupControllerV3Interface";
-import { NoteLookupProviderUtils } from "../components/lookup/utils";
+import { NoteLookupProviderUtils } from "../components/lookup/NoteLookupProviderUtils";
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/ConvertLink.ts:25]: as a rule of thumb, when we have a lot of utils in a folder, can we split it out into an utils folder?

```
/lookup/utils/index.ts
/lookup/utils/NoteLookupProvider.ts
...
```

don't need to change for this one but for future PRs, lets stick to amoeba in our code base as well

File: packages/plugin-core/src/test/MockDendronExtension.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        "dendron.rootDir": string;
+        "dendron.dendronDir": string;
+        "dendron.logLevel": string;
+        "dendron.trace.server": string;
+        "dendron.serverPort": string;
+      }>
+    | undefined {
+    throw new Error("Method not implemented in MockDendronExtension.");
+  }
+  getWorkspaceSettingOrDefault({}: {
```
Comment (kevinslin) [inline packages/plugin-core/src/test/MockDendronExtension.ts:134]: can you make sure `eslint` is enabled in your workspace by downloading and enabling the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

**Pull request merged**